pkg/plugin/keyvault.go

```diff
@@ -22,33 +25,62 @@ import (
 	"github.com/Azure/go-autorest/autorest"
 	"github.com/Azure/go-autorest/autorest/azure"
 	"k8s.io/klog/v2"
+	"k8s.io/kms/pkg/service"
+)
+
+// encryptionResponseVersion is validated prior to decryption.
+// This is helpful in case we want to change anything about the data we send in the future.
+var encryptionResponseVersion = "1"
+
+const (
+	dateAnnotationKey             = "date.azure.akv.io"
+	requestIDAnnotationKey        = "x-ms-request-id.azure.akv.io"
+	keyvaultRegionAnnotationKey   = "x-ms-keyvault-region.azure.akv.io"
+	versionAnnotationKey          = "version.azure.akv.io"
+	algorithmAnnotationKey        = "algorithm.azure.akv.io"
+	dateAnnotationValue           = "Date"
+	requestIDAnnotationValue      = "X-Ms-Request-Id"
+	keyvaultRegionAnnotationValue = "X-Ms-Keyvault-Region"
 )
 
 // Client interface for interacting with Keyvault.
 type Client interface {
-	Encrypt(ctx context.Context, cipher []byte) ([]byte, error)
-	Decrypt(ctx context.Context, plain []byte) ([]byte, error)
+	Encrypt(
+		ctx context.Context,
+		plain []byte,
+		encryptionAlgorithm kv.JSONWebKeyEncryptionAlgorithm,
+	) (*service.EncryptResponse, error)
+	Decrypt(
+		ctx context.Context,
+		cipher []byte,
+		encryptionAlgorithm kv.JSONWebKeyEncryptionAlgorithm,
+		apiVersion string,
+		annotations map[string][]byte,
+		decryptRequestKeyID string,
+	) ([]byte, error)
 }
 
-type keyVaultClient struct {
+// KeyVaultClient is a client for interacting with Keyvault.
+type KeyVaultClient struct {
 	baseClient       kv.BaseClient
 	config           *config.AzureConfig
 	vaultName        string
 	keyName          string
 	keyVersion       string
 	vaultURL         string
+	keyID            string
 	azureEnvironment *azure.Environment
 }
 
 // NewKeyVaultClient returns a new key vault client to use for kms operations.
-func newKeyVaultClient(
+func NewKeyVaultClient(
 	config *config.AzureConfig,
 	vaultName, keyName, keyVersion string,
 	proxyMode bool,
 	proxyAddress string,
 	proxyPort int,
 	managedHSM bool,
-) (*keyVaultClient, error) {
+) (*KeyVaultClient, error) {
```

Comment:
```suggestion
) (Client, error) {
```

could this just return the `Client` interface?